src/main/java/com/fdmgroup/forex/services/FxRateService.java

```diff
@@ -1,38 +1,135 @@
+            return updateFxRatesFromUSDRates(usdRates);
+        } catch (Exception e) {
+            throw new InternalServerErrorException("Error updating exchange rates via external API: " + e.getMessage());
+        }
+    }
+
+    private String fetchFxRatesFromAPI() throws IOException {
+        URL obj = new URL(fxApiUsdUrl);
+        HttpURLConnection con = (HttpURLConnection) obj.openConnection();
+        con.setRequestMethod("GET");
+        int responseCode = con.getResponseCode();
+
+        if (responseCode == HttpURLConnection.HTTP_OK) {
+            BufferedReader input = new BufferedReader(new InputStreamReader(con.getInputStream()));
+            StringBuilder response = new StringBuilder();
+            String inputLine;
+            while ((inputLine = input.readLine()) != null) {
+                response.append(inputLine);
+            }
+            input.close();
+            return response.toString();
+        } else {
+            throw new IOException("Failed to fetch exchange rates from API");
+        }
+    }
+
+    private Map<String,Double> parseFxRatesResponse(String response) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        Map<String, Object> data = mapper.readValue(response, new TypeReference<Map<String, Object>>() {});
+
+        Object ratesObject = data.get("rates");
+        if (!(ratesObject instanceof Map<?, ?>)) {
+            throw new IllegalArgumentException("Expected 'rates' to be a Map, but got: " + ratesObject.getClass());
+        }
+    
+        Map<?, ?> rawRates = (Map<?, ?>) ratesObject;
+        Map<String, Double> rates = new HashMap<>();
+    
+        for (Map.Entry<?, ?> entry : rawRates.entrySet()) {
+            if (!(entry.getKey() instanceof String)) {
+                throw new IllegalArgumentException("Expected rate key to be a String, but got: " + entry.getKey().getClass());
+            }
+            String key = (String) entry.getKey();
+            Object value = entry.getValue();
+    
+            if (value instanceof Integer intValue) {
+                rates.put(key, intValue.doubleValue());
+            } else if (value instanceof Double doubleValue) {
+                rates.put(key, doubleValue);
+            } else if (value instanceof Number numberValue) {
+                rates.put(key, numberValue.doubleValue());
+            } else {
+                throw new IllegalArgumentException("Unexpected value type in rates map: " + value.getClass());
+            }
+        }
+        return rates;
+    }
+
+    private List<FxRate> updateFxRatesFromUSDRates(Map<String,Double> rates) {
+        List<FxRate> fxRates = new ArrayList<>();
+        for (String code : rates.keySet()) {
+            double rate = Double.valueOf(rates.get(code));
+            try {
+                FxRate fxRate = createOrUpdateFxRate(code, rate);
+                fxRates.add(fxRate);
+            } catch (RecordNotFoundException e) {
```

Comment:
`RecordNotFoundException` already caught in `createOrUpdateFxRate` method

Reply:
These serve different purposes: in the `createOrUpdateFxRate` method the try-catch is used to check whether the `FxRate` already exists. If it does not, it will throw the exception when we try to find it, so then we grab the relevant currency and create the new `FxRate` in the `catch` block.

The try-catch on the parent method is for the edge case in which the `currency` also does not exist and so raises the same type of error when we try to find it in the currency repo - this is unlikely since the exchange rates API should return only valid currency codes that we are already aware of, and we will also initialise our currency objects at the start of the application before initialising exchange rates.

It isn't well documented the way I currently have it as an empty catch block, and perhaps it is unnecessary if we aren't concerned about the above edge case.

The main point is that if there is an issue with one fx rate being passed in, we don't want it to block the updates to the remaining valid fx rates.

Reply:
Ah I can see that now. In that case we better keep it to handle the edge case.